CorfuDB/CorfuDB
# Preserve Transaction Boundary on LR Receiver.

## Overview
During log entry sync, multiple transactions are batched in a single message by the Sender.  Before this fix, the receiver coalesced and applied them all as a single  transaction.  This PR addresses the issue and maintains the transaction boundaries to be the same as what was received.

Why should this be merged: 
As multiple transactions are grouped in a single transaction on the receiver, the order of streaming updates is not consistent with the sender without this fix.

Tests Added
--------------

LogEntryWriterTest
---------------------
1. Verify that each opaque entry in a LogReplicationEntryMsg is applied
   in a separate transaction.
2. Verify that redundant apply of the same LogReplicationEntryMsg is not
   performed.
3. Verify that 2 LogReplicationEntryMsgs with overlapping sequence
   numbers are applied correctly
4. Simulate an exception after a LogReplicationEntryMsg is partially
   applied.  Verify that a subsequent resend of this message is correctly applied.
5. Verify that a LogReplicationEntryMsg with a mismatching 'previous
   Timestamp' is rejected.
6. Verify that a LogReplicationEntryMsg with a mismatching 'snapshot
   timestamp' (lower and higher) is not applied.
7. Verify that a LogReplicationEntryMsg with a mismatching topology id
   is rejected.

MetadataManagerTest
-------------------------
1. Verify that metadata is correctly updated after a successful LogEntry
   sync.
2. Simulate concurrent update of topology id and log entry sync.  Verify
   that the metadata is updated according to which transaction(topology
   update or log entry apply) finished first.
3. Verify that MetadataManager returns Address.NON_ADDRESS as the
   initial/default value of 'LAST_SNAPSHOT_APPLIED' and
   'LAST_LOG_ENTRY_BATCH_PROCESSED'.

CorfuStoreShimTest
-------------------------
1. Verify that for a WriteAfterWrite TX type, touch() api is needed to detect
   a conflict between a read-only and write transaction on a conflict key.
   Also verify that the conflict is not detected without the touch() api.



Related issue(s) (if applicable): #<number>


## Checklist (Definition of Done):

- [ ] There are no TODOs left in the code
- [ ] [Coding conventions](https://github.com/CorfuDB/CorfuDB/wiki/Corfu-Style-Guidelines) (e.g. for logging, unit tests) have been followed
- [ ] Change is covered by automated tests
- [ ] Public API has Javadoc
